### ISSUE TITLE
[bug] Error not thrown when detected compiler does not match profile

### ISSUE THREAD

**Issue by samuel-emrys** [conan-io/conan#14755]
### Environment details

* Operating System+version: Arch Linux
* Compiler+version: clang 16.0.6
* Conan version: 2.0.10
* Python version: 3.11.1

### Steps to reproduce

1. Install clang 16.0.6
2. Set default conan profile to the following, noting that `compiler.version=15`
```
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=clang
compiler.cppstd=20
compiler.libcxx=libstdc++11
compiler.version=15
os=Linux
[conf]
tools.build:compiler_executables={'c': '/usr/bin/clang', 'cpp': '/usr/bin/clang++'}

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=clang
compiler.cppstd=20
compiler.libcxx=libstdc++11
compiler.version=15
os=Linux
[conf]
tools.build:compiler_executables={'c': '/usr/bin/clang', 'cpp': '/usr/bin/clang++'}

```
4. `conan new cmake_lib -d name=test -d version=0.1.0`
5. `conan build .`
6. Observe no error is thrown. This always used to be the case.

### Logs

```
$ conan --version
Conan version 2.0.10
$ conan new cmake_lib -d name=test -d version=0.1.0
File saved: CMakeLists.txt
File saved: conanfile.py
File saved: include/test.h
File saved: src/test.cpp
File saved: test_package/CMakeLists.txt
File saved: test_package/conanfile.py
File saved: test_package/src/example.cpp
$ clang++ --version
clang version 16.0.6
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
$ which clang++
/usr/bin/clang++
$ conan build .

======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=clang
compiler.cppstd=20
compiler.libcxx=libstdc++11
compiler.version=15
os=Linux
[conf]
tools.build:compiler_executables={'c': '/usr/bin/clang', 'cpp': '/usr/bin/clang++'}

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=clang
compiler.cppstd=20
compiler.libcxx=libstdc++11
compiler.version=15
os=Linux
[conf]
tools.build:compiler_executables={'c': '/usr/bin/clang', 'cpp': '/usr/bin/clang++'}


======== Computing dependency graph ========
Graph root
    conanfile.py (test/0.1.0): /home/user/projects/test/conan-compiler-error-check/conanfile.py

======== Computing necessary packages ========

======== Installing packages ========

======== Installing packages ========

======== Finalizing install (deploy, generators) ========
conanfile.py (test/0.1.0): Calling generate()
conanfile.py (test/0.1.0): Generators folder: /home/user/projects/test/conan-compiler-error-check/build/Release/generators
conanfile.py (test/0.1.0): CMakeToolchain generated: conan_toolchain.cmake
conanfile.py (test/0.1.0): Preset 'conan-release' added to CMakePresets.json. Invoke it manually using 'cmake --preset conan-release' if using CMake>=3.23
conanfile.py (test/0.1.0): If your CMake version is not compatible with CMakePresets (<3.23) call cmake like: 'cmake <path> -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=/home/user/projects/test/conan-compiler-error-check/build/Release/generators/conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release'
conanfile.py (test/0.1.0): CMakeToolchain generated: CMakePresets.json
conanfile.py (test/0.1.0): CMakeToolchain generated: ../../../CMakeUserPresets.json
conanfile.py (test/0.1.0): Generating aggregated env files
conanfile.py (test/0.1.0): Generated aggregated env files: ['conanbuild.sh', 'conanrun.sh']

======== Calling build() ========
conanfile.py (test/0.1.0): Calling build()
conanfile.py (test/0.1.0): Running CMake.configure()
conanfile.py (test/0.1.0): RUN: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/home/user/projects/test/conan-compiler-error-check/build/Release/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/home/user/projects/test/conan-compiler-error-check" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/home/user/projects/test/conan-compiler-error-check"
-- Using Conan toolchain: /home/user/projects/test/conan-compiler-error-check/build/Release/generators/conan_toolchain.cmake
-- Conan toolchain: Setting CMAKE_POSITION_INDEPENDENT_CODE=ON (options.fPIC)
-- Conan toolchain: C++ Standard 20 with extensions OFF
-- Conan toolchain: Setting BUILD_SHARED_LIBS = OFF
-- The CXX compiler identification is Clang 16.0.6
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done (0.3s)
-- Generating done (0.0s)
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_POLICY_DEFAULT_CMP0091


-- Build files have been written to: /home/user/projects/test/conan-compiler-error-check/build/Release

conanfile.py (test/0.1.0): Running CMake.build()
conanfile.py (test/0.1.0): RUN: cmake --build "/home/user/projects/test/conan-compiler-error-check/build/Release" -- -j24
[ 50%] Building CXX object CMakeFiles/test.dir/src/test.cpp.o
[100%] Linking CXX static library libtest.a
[100%] Built target test

```

**Comment by memsharded** [comment]
Hi @samuel-emrys 

Thanks for your feedback.

Automatic compiler check was something that the legacy ``cmake`` generator implemented. 
It was so problematic and giving a bunch of false positives blocking users, that we added a explicit CMake variable to skip this compiler check, that users could define in their ``CMakeLists.txt``, because the ``cmake`` generator was intrusive anyway. Besides the ``cmake`` generator, no other build system integration actually checked the compiler matching the profile, because other build systems lacks the capabilities to implement this properly. It has always been considered the responsibility of the users to have a profile that matches their actual compiler.

As that detection within ``cmake`` generator was not very reliable, it has not been moved forward yet to the new generators like ``CMakeToolchain`` (this responsibility would belong to it). That doesn't mean that it cannot be added to it, marking it as a "feature", but I am afraid at the moment this won't be a priority, as it is not trivial, and still can have issues, so it would be added as opt-in in the best case.

Another possibility is to expand the ``profile.py`` plugin to actually do compiler checks that could benefit all build systems. These checks need to be fast, otherwise it can be inconvenient. But also these checks are complicated to be robust, it is almost impossible to know how the build systems will pick their compilers. This means that a general ``profile.py`` implementation seems unlikely, but users could do their own extensions for their setups that they know very well and can define some pre-conditions.

**Comment by samuel-emrys** [comment]
ah, no problems :) If this is intended behaviour, disregard. I guess I've only recently started noticing this after transitioning to conan 2, so I assumed that behaviour hadn't changed.

For what it's worth, I found it a useful check and was an early failure indicating that I was doing something that I didn't intend to do. I didn't find the ux difficult to navigate and didn't experience any false positives. Having said that, I can see how it might add friction to the new user experience as my observation is that new users don't understand the role that profiles play straight away and rely heavily on the auto detection.

P.S, i think the `cmake_find_package` generator may also have checked? I never really used the `cmake` generator, so I'm not sure why I would have seen the errors unless I was using it accidentally/in test packages maybe.

Happy for you to close this as this is not a bug, otherwise happy for this to be turned into a feature request if you think that's appropriate.

**Comment by memsharded** [comment]
> ah, no problems :) If this is intended behaviour, disregard. I guess I've only recently started noticing this after transitioning to conan 2, so I assumed that behaviour hadn't changed.

Yeah, it is already this way in Conan 1.X new CMakeDeps/CMakeToolchain generator, not specific of Conan 2

> For what it's worth, I found it a useful check and was an early failure indicating that I was doing something that I didn't intend to do. I didn't find the ux difficult to navigate and didn't experience any false positives. Having said that, I can see how it might add friction to the new user experience as my observation is that new users don't understand the role that profiles play straight away and rely heavily on the auto detection.

Yes, it was indeed an useful check, I am not saying it was not, just that the false positives were a bad experience, enough annoying to drop it in the new generators. I think most of the problems came for special compilers, like cross-compilers, proprietary special compilers, anything that deviates a bit from the standard is likely to cause issues.

> P.S, i think the cmake_find_package generator may also have checked? I never really used the cmake generator, so I'm not sure why I would have seen the errors unless I was using it accidentally/in test packages maybe.

``cmake_find_package`` was used in conjunction a lot with ``cmake``, because it didn't contain the "toolchain" part, so that could be the reason.

> Happy for you to close this as this is not a bug, otherwise happy for this to be turned into a feature request if you think that's appropriate.

Lets keep it as a feature request, as said above, it is not that it doesn't have value, and I have known of other users that have missed it, and they thought it was useful too

**Comment by OleksandrKvl** [comment]
I have 2 scenarios where I'd like to have this feature in some way:
- on macOS I have the most recent Clang version installed via brew. I used this path in my `clang-brew` profile `/opt/homebrew/opt/llvm/bin/clang` (notice, it's not `clang-19`). And when clang gets a major version update, this profile keeps working, although it's not correct. One solution for this case is to use `clang-19` binary name but having an error from Conan would be nice.
- on GitHub CI different images can have different compiler versions installed by default. It would be nice to know that compiler version specified in profile is guaranteed to be used. And for MSVC it's not possible to use `compiler-VERSION` name so no way to ensure specific version.

**Comment by memsharded** [comment]
Thanks for the feedback.

Still, the problem with this ticket is that it is basically impossible to do these checks and have at the same time the transparent (and CMake recommended) integrations via CMake toolchains, and other possible approaches (like auto-detection in profile plugin) have proven to be fragile and error prone, with annoying false positives too.

**Comment by OleksandrKvl** [comment]
I'm sure you have more experience and understanding of this feature but why can't it be just an option (OFF by default) either in profile or in CLI? Like:

```
# profile configuration
[settings]
compiler.validate_version=true
```

I don't understand why CMake or any other generators should be involved here since this check should happen before their functionality is needed.

**Comment by memsharded** [comment]
> I don't understand why CMake or any other generators should be involved here since this check should happen before their functionality is needed.

Because the important misalignment happens when CMake uses a different compiler than the declared one. The check that the OP was missing was a check in the old CMake generators in Conan 1.X. Because the rules how CMake picks the compiler are a bit more complex than what Conan can detect, and it depends on a lots of factors, from the provided user Cmake toolchains, to possible hardcoded definitions in the user CMakeLists.txt. So the only possible way to check those mismatches is within CMake.


For other checks, there is basically no need from Conan, as Conan already provides a ``detect_api`` that you can use in your profiles to define the Conan auto-detected compiler, compiler-version, arch, etc.

**Comment by OleksandrKvl** [comment]
> Conan already provides a detect_api that you can use in your profiles to define the Conan auto-detected compiler, compiler-version, arch, etc.

How can I use `detect_api` for checks? Seems that its main purpose is to generate profiles for a specific compiler vendor.

**Comment by memsharded** [comment]
The thing is that if you want to use the compiler that Conan can identify as the "active" one, no need to check. Just use either an auto-detected profile (``conan profile detect``) or use the ``detect_api`` in your own profile to detect the compiler following Conan detection.

If you really want to check compiler instead of just defining it, the profiles support jinja templating (https://docs.conan.io/2/reference/config_files/profiles.html#profile-rendering) so something in the line (pseudo code, not tested):
```
{% set compiler, version, compiler_exe = detect_api.detect_default_compiler() %}
{% if version != 9 %}
error
{% else %}
compiler.version=9
{% endif %}


The problem is that those checks don't really guarantee that CMake will pick that compiler, because of the above reasons, and this is why the OP was requesting to try to add checks in CMake generators.

**Comment by OleksandrKvl** [comment]
I googled how to trigger an error from Jinja template and found nothing, your approach is simpler, thanks :)

**Comment by memsharded** [comment]
Yeah, as the evaluation of the profile will fail for not expected inputs, you can just render there the message string in the profile and it will be in the final error message.
